torch/csrc/distributed/rpc/agent_utils.cpp

```diff
@@ -40,6 +40,93 @@ std::unordered_map<std::string, worker_id_t> collectNames(
   return nameToId;
 }
 
+const string barrierKeyPlaceholder = "_BARRIER_";
+const string barrierReady = "_READY";
+static std::atomic<int> barrierId(0);
+
+// A mutex and a cv to guard access to the call counts and watch for changes.
+std::mutex barrierMutex_;
+
+string getNextBarrierKey() {
+  barrierId++;
+  return barrierKeyPlaceholder + std::to_string(barrierId);
+}
+
+// Parse the key value to return, split based on "_"
+std::tuple<int, int> parseKeyValue(string keyValue) {
+  std::string delimiter = "_";
+  int numProcess = std::stoi(keyValue.substr(0, keyValue.find(delimiter)));
+  int totalActiveCalls =
+      std::stoi(keyValue.substr(keyValue.find(delimiter) + 1));
+  return std::make_tuple(numProcess, totalActiveCalls);
+}
+
+// Creates a key value of form "{process_num}_{total_active_client_calls}"
+std::string createKeyValue(int nextProcessNum, int totalActiveCalls) {
+  return std::to_string(nextProcessNum) + "_" +
+      std::to_string(totalActiveCalls);
+}
+
+// Performs a barrier with store
+// An optional active_count argument can be used to determine the number
+// of total active client calls among all agents
+// Returns true if we are checking for calls and there are 0 active calls
+// Returns false otherwise
+bool barrier(
+    ::c10d::PrefixStore store,
+    const int worldSize,
+    bool checkCalls,
+    int activeCalls) {
+  std::unique_lock<std::mutex> lock(barrierMutex_);
+  VLOG(1) << "Starting store-based barrier";
+  std::string barrierKey = getNextBarrierKey();
+  bool keyCreated = false;
+  std::string currKeyValue, newKeyValue, desiredValue;
+  int processNum = 0;
+  int totalActiveCalls = 0;
+  while (true) {
+    // Check if key exists
+    keyCreated = store.check(std::vector<std::string>{barrierKey});
+
+    if (!keyCreated) {
+      currKeyValue = "";
+      processNum = 0;
+      desiredValue = createKeyValue(processNum + 1, activeCalls);
+    } else {
+      std::vector<uint8_t> valueArr = store.get(barrierKey);
+      currKeyValue = std::string(valueArr.begin(), valueArr.end());
+      std::tie(processNum, totalActiveCalls) = parseKeyValue(currKeyValue);
+      desiredValue =
+          createKeyValue(processNum + 1, totalActiveCalls + activeCalls);
```

Comment:
I guess this part is the race condition as it's possible that multiple processes are reading the same `totalActiveCalls` and then compareSet won't sum the together?

Reply:
I wonder if we should use `add()` instead?

Reply:
just wanna share a bit more about the idea, assume `add()` is atomic. We can have two keys here, one for the number reported processes, and another for the number of total active calls. Each Process always first update the total active calls, and then update the total number of reported processes. The last process (i.e., the one that sees the total number of reported processes equals the world size) is responsible to set the ready key and every process should then read the total number of active calls on their own. I assume this can also avoid using the while loop? There might be other ways to further simplify it.